### ISSUE TITLE
Auto-merge support with "restrict who can push to this branch"

### ISSUE THREAD

**Issue by nesl247** [dependabot/feedback#86]
Currently auto-merge is not possible for any branch that has "restrict who can push to this branch" enabled in branch protections.

Unfortunately GitHub does not currently allow adding bots to the list.

Is there any way to communicate with GitHub, especially since Dependabot is a Marketplace app?

**Comment by greysteil** [comment]
There's an issue on the GitHub platform forum, which is their preferred way to receive app-related feedback. Adding your voice there might help:

https://platform.github.community/t/repositories-which-have-protected-branches-with-push-restrictions-have-no-ability-to-grant-push-rights-to-integrations/1376/48

**Comment by nesl247** [comment]
I've gone ahead and replied and asked for an update. Unfortunately GitHub is slow as can be to do the most basic of things.

**Comment by kimadactyl** [comment]
Plus one, is it possible for us to add `dependabot-bot` as a user on the protected branch? It would require the bot responding to "add to team" requests. That's how Rultor does it :)

**Comment by greysteil** [comment]
Good question. Sadly not at the moment - we'd need to build a bunch of automation around `dependabot-bot` to make that work. It doesn't sound *too* hard, though, and is definitely the best workaround I've heard so far. I'm a little underwater with other requests, but will try to come back to this - ping me if I don't (and if there's no progress from GitHub - we're still pushing them).

**Comment by cihati** [comment]
Any progress on this one @greysteil ?

**Comment by greysteil** [comment]
This one's still with GitHub support. I chased them about it last week and just got another "it's on our roadmap" email. I know they're very busy, but hope this will become a priority eventually. In the meantime, the best way to help is probably emailing support@github.com.

**Comment by simlu** [comment]
We'd be very interested in a proper solution as well... Basically we want human prs to be reviewed before merging into our dev branch. However dependabot prs should be ok to auto merge. We could solve this independently from dependabot, however ideally it's solved by dependabot (if possible).

@greysteil What would be a appropriate request for the email you provided? If you post a template here more people might be tempted to sent that.

**Comment by greysteil** [comment]
I'd go for something like this:

To support@github.com:
> Hi team,
>
> I've got a feature request: we use Dependabot to create pull requests on **repo-name** and would like it to be able to automatically merge the PRs it creates, whilst continuing to require pull request reviews for PRs created by humans.
>
> Currently, it looks like that's not possible under the branch-protection rules, particularly as Dependabot acts as a GitHub app (its username is dependabot[bot]). Are there any plans to add support?
>
> Many thanks,
>
> **your-name**

**Comment by bennyn** [comment]
@greysteil According to [this post](https://github.community/t5/How-to-use-Git-and-GitHub/Where-do-I-post-feature-requests-for-Github/m-p/15027#M4671) we can bring up feature requests on https://github.community. How about creating a ticket there? We can then upvote and support it. 😃

**Comment by simlu** [comment]
@greysteil Thanks!

We've solved this now in circleci using gally and a bot github user. Not as clean as I'd like it to be (oh circle why do you make our life so hard), but works great.

Excerpt (job):

```yml
  gally-auto-merge:
    docker:
      - image: 'circleci/node:10'
    steps:
      - run:
          name: Set PR number because circle :(
          command: |
            echo 'export CIRCLE_PR_NUMBER="${CIRCLE_PR_NUMBER:-${CIRCLE_PULL_REQUEST##*/}}"' >> $BASH_ENV
            source $BASH_ENV
            echo $CIRCLE_PR_NUMBER
      - checkout
      - run: sudo npm i -g gally
      - run: ga approve $CIRCLE_PR_NUMBER --condition "base.ref=dev&state=open&user.login=dependabot[bot]"
      - run: ga merge $CIRCLE_PR_NUMBER --condition "base.ref=dev&state=open&user.login=dependabot[bot]"
```

If you don't need approval you can skip that step in the job.

**Comment by jodastephen** [comment]
I've played with [this bot](https://github.com/dkhmelenko/autoapproval), and with some hacking managed to get it to auto-approve dependabot PRs so they then can be merged.

@greysteil, Ideally, this would be a new config option in dependabot, but I'm not sure if it is as simple as that. You can't normally approve your own PR, but maybe you can through the API? Otherwise, how about a new bot - approvabot ?

**Comment by greysteil** [comment]
We can't approve our own PRs, but a GitHub action to automatically approve them ought to work. I think @hmarr had something basic thrown together.

**Comment by hmarr** [comment]
I've just fixed up the auto-approval action I threw together a while ago and made the repo open source.

If you have access to GitHub Actions (and are using a private repo, as I gather they're not yet enabled for open source repos), this should do the trick for you.

https://github.com/hmarr/actions/#auto-approve

**Comment by Mcat12** [comment]
@hmarr That bot does not seem to solve the issue.

For clarification, this issue is about the "Restrict who can push to matching branches" option, which only allows the specified users or teams to push to the branch. What your Action solves for is when "Require pull request reviews before merging" is enabled, which is not the point of this issue.

**Comment by nicolasrouanne** [comment]
Any workaround **_without_ GitHub Actions** (I'm not yet into the beta 😢)?

**Comment by greysteil** [comment]
We're working on a general fix for this in GitHub right now (i.e., allowing you to specify bots as actors that are allowed to push to protected branches).

**Comment by nicolasrouanne** [comment]
For those who don't have GitHub Actions yet the **`autoapproval` bot** is working great. You can _self host it_ ([source code](https://github.com/dkhmelenko/autoapproval)) or _add the [GitHub app](https://github.com/apps/autoapproval)_ (that's what I did).

Here is my `.github/autoapproval.yml` settings file:
```yml
from_owner:
  - dependabot-preview[bot]
required_labels:
  - dependencies
apply_labels:
  - autoapproved
```

<img width="663" alt="Screenshot 2019-06-24 at 11 15 16" src="https://user-images.githubusercontent.com/12157019/60007241-36b85900-9672-11e9-9708-8b6209da59bb.png">

So basically:
- `autoapproval` approves _dependabot_ PRs automatically
- `dependabot` auto-merges its PRs since they are approved
- me: ☕️

**Comment by Mcat12** [comment]
@nicolasrouanne that kind of fix is unrelated to this issue, as I previously mentioned:
> For clarification, this issue is about the "Restrict who can push to matching branches" option, which only allows the specified users or teams to push to the branch. What your Action solves for is when "Require pull request reviews before merging" is enabled, which is not the point of this issue.

https://github.com/dependabot/feedback/issues/86#issuecomment-495785587

**Comment by greysteil** [comment]
@Mcat12 we'll have a fix for that direct in GitHub (and usable for other apps, not just Dependabot) in the next few days. @feelepxyz is working on it right now 🙂

**Comment by Mcat12** [comment]
@greysteil Can you update this issue when the feature has been released?

**Comment by greysteil** [comment]
For sure. The whole of GitHub is at a big offsite this week so it won't happen in the next 7 days, but the code is almost ready and should be deployed next week.

**Comment by karltaylor** [comment]
@greysteil enjoy your offsite

**Comment by StephanBijzitter** [comment]
Perhaps it should be an issue of its own, but for several of our repositories we do not trust our test coverage enough to go full automatic.

One of our repositories now uses autoapprove + dependabot and that's working nicely, but for two other repositories we'd prefer to manually approve pull requests from dependabot and have dependabot merge them as soon as all other status checks have come back positive. That way we don't have to wait for the build to complete.

I may have been rambling a bit, so maybe this could summarise the option I'd also like to see (next to the fully automatic solution):
- Dependabot creates PR
- Developer approves while build is in progress
- Build finishes
- Dependabot merges PR

And

- Dependabot creates PR
- Build finishes
- Developer approves
- Dependabot merges PR

Currently we get:
> Dependabot tried to automerge this PR, but received the following error from GitHub:

> At least 1 approving review is required by reviewers with write access.

> As a result, we've disabled automerging on this repo (you can re-enable it in your Dependabot settings).

Which is preventing us from doing that while we're working on our test coverage

**Comment by greysteil** [comment]
@StephanBijzitter in your account settings there's an option called "Treat PR approval as a request to merge". If you switch that on then you'll get the flow you described above.

**Comment by StephanBijzitter** [comment]
You guys are the best

**Comment by philipbjorge** [comment]
@greysteil -- Is there an update on this?
Mostly wondering if there is some flag/checkbox I need to mark somewhere because this still isn't working for us. Thanks 😄!

**Comment by greysteil** [comment]
Not yet but there should be soon - we’re waiting for review internally but this is a complicated change to permissioning

**Comment by acazacu** [comment]
Are there any updates on the incoming change?

**Comment by greysteil** [comment]
It’s still happening, but it’s slow. Here’s how it works internally at GitHub:
- Dependabot is a team here
- Another team (ecosystem-apps) owns the experience and code here
- We’ve chatted to them, got the work scopes, and started working on it with their blessing (they’ll be responsible for hitting merge)
- As part of the work we’ve found a bunch of debt around permissions which we’re working to clear up, but it’s slow going for us because it’s not our area and the person who was working on it is on holiday
- We’ll keep pushing and get this over the line (I’ll update here if we ever *don’t* have someone working on it!)

Stay with us - we’re working on it but it’s hard!

**Comment by NargiT** [comment]
You could use policy-bot from palantir that replace and enhance the branch protection by adding a status check to the pull request. If you configure your repo with branch protection for status check. Then you should be able to merge even if it is a bot. https://github.com/palantir/policy-bot

**Comment by feelepxyz** [comment]
👋 We've just shipped the UI to allow installed GitHub Apps to push to protected branches, you can add installed apps from the repository settings for the protected branch:

<img width="1125" alt="Screenshot 2019-08-22 at 11 30 40" src="https://user-images.githubusercontent.com/20165/63507724-4a483a80-c4d0-11e9-87c2-54f2a2af071c.png">

Short video walkthrough: https://drive.google.com/open?id=1i15psbTl7LhSMhNCi9tom-OdgybM9slQ

If you've got required pull request reviews set Dependabot will still fail to merge the pull request until it's approved. You can get around this with a GitHub action to auto approve these pull requests: https://github.com/hmarr/auto-approve-action

We're still working on the corresponding APIs to query/update apps as allowed actors on protected branches (GraphQL v4 and Rest v3). Planning on publishing these by next week.

There's also a bug when using git auth (e.g. git push with credentials) preventing apps from pushing to protected branches even though it's been authorized. Should be fixed by next week.

🎉

**Comment by mark-edf** [comment]
does the API for branch protection accept `dependabot-preview` too as a user or will that come later

**Comment by feelepxyz** [comment]
@simlu the APIs are now live 🎉 

Updated GraphQL API:

- https://developer.github.com/v4/object/pushallowance/
- https://developer.github.com/v4/changelog/2019-08-23-schema-changes/

Updated Rest API: https://developer.github.com/changes/2019-09-05-apps-protected-branches-api/

**Comment by localheinz** [comment]
Any news on this one?

**Comment by feelepxyz** [comment]
@localheinz this should be working now if you add `Dependabot Preview` to the list of actors with push access on the protected branch. What's not working?

**Comment by localheinz** [comment]
@feelepxyz 

The actor `Dependabot Preview` (or any other `Dependabot` actor) does not show up:

![Screen Shot 2019-12-16 at 11 38 36](https://user-images.githubusercontent.com/605483/70900214-9fb4d400-1ff8-11ea-89cc-e08830ad5616.png)

**Comment by philsturgeon** [comment]
I'm also not seeing the name show up.

<img width="575" alt="Screen Shot 2019-12-16 at 11 49 49" src="https://user-images.githubusercontent.com/67381/70900993-3766f200-1ffa-11ea-813a-871b5f72d919.png">

**Comment by feelepxyz** [comment]
@philsturgeon @localheinz oh looks like this is broken. Will investigate!

**Comment by feelepxyz** [comment]
@localheinz @philsturgeon workong on a fix in github but going to take some time to get this out amidst the holidays. You can work around it by explicitly selecting the repos that `Dependabot Preview` has access to instad of `All repositories` on the installation.

**Comment by localheinz** [comment]
@feelepxyz 

Can confirm that selecting repositories manually allows to select `@dependabot-preview`!

<img width="699" alt="Screen Shot 2019-12-21 at 09 52 16" src="https://user-images.githubusercontent.com/605483/71305667-a7e58880-23d7-11ea-92fb-e5cebb55e24b.png">

**Comment by localheinz** [comment]
@feelepxyz 

Not sure if this is a _different_ issue, but I still do not see Dependabot automatically merge pull requests.

### Apps allowed to push to matching branches

`@dependabot-preview` is on the list of apps allowed to push to matching branches (in this case, `master`)

![Screen Shot 2019-12-22 at 16 45 10](https://user-images.githubusercontent.com/605483/71324055-8caf7180-24da-11ea-94b5-d93f029b3e1a.png)

### `.dependabot/config.yml`

[`.dependabot/config.yml`](https://github.com/ergebnis/composer-normalize/blob/54cd5fc8a76b1f1c211fbd07bdcc330294949b6a/.dependabot/config.yml) hopefully has the appropriate configuration:

```yml
# https://dependabot.com/docs/config-file/

version: 1

update_configs:
  - automerged_updates:
      - match:
          dependency_type: "development"
    default_assignees:
      - "localheinz"
    default_labels:
      - "dependency"
    default_reviewers:
      - "localheinz"
    directory: "/"
    ignored_updates:
      - match:
          dependency_name: "composer-plugin-api"
      - match:
          dependency_name: "composer/composer"
    package_manager: "php:composer"
    update_schedule: "live"
    version_requirement_updates: "increase_versions"
```

### `CODEOWNERS`

I also have [`CODEOWNERS`](https://github.com/ergebnis/.github/blob/6b414d9aa67edea91b77b5910594103483ed922a/CODEOWNERS) configured on the organization level.

### Example

I would have expected https://github.com/ergebnis/composer-normalize/pull/283 to be merged automatically.

However, this has not happened.

### Modifications required?

Do I need to remove the `default_reviewers` configuration?

```diff
diff --git a/.dependabot/config.yml b/.dependabot/config.yml
index 26a20fe..0778a1b 100644
--- a/.dependabot/config.yml
+++ b/.dependabot/config.yml
@@ -10,8 +10,6 @@ update_configs:
       - "localheinz"
     default_labels:
       - "dependency"
-    default_reviewers:
-      - "localheinz"
     directory: "/"
     ignored_updates:
       - match:
```

Should I stop using `CODEOWNERS` or configure it differently to exclude `composer.json` and `composer.lock`?

**Comment by romrell4** [comment]
I'm still not seeing the option to add dependabot-preview (or any other dependabot users) to my list of people with push access. Is this bug not completely fixed?

**Comment by feelepxyz** [comment]
@romrell4 this should be fixed now. Was on hold over the break. 

@localheinz yeah looks like the required reviewers settings are preventing the automerge: https://github.com/ergebnis/composer-normalize/pull/283#issuecomment-568273729 there's an option to treat a PR approval as a request to merge in your [dependabot dashboard](https://app.dependabot.com/) settings: 

<img width="602" alt="Screen Shot 2020-01-06 at 12 48 00" src="https://user-images.githubusercontent.com/20165/71819105-d7a4b980-3082-11ea-9cb2-fedb6de57e83.png">

**Comment by localheinz** [comment]
Thank you, @feelepxyz!

**Comment by chris48s** [comment]
Just to clarify my understanding on this issue, is this summary of the situation correct?

Dependabot _can_ now be added with permissions to push to protected branches, but it _can't_ approve its own PRs. In order to trigger the merge, you still need either a manual approval, or another bot/action (other than dependabot) to approve the PR, even if a PR matches rules in [automerged_updates](https://dependabot.com/docs/config-file/#automerged_updates).

**Comment by feelepxyz** [comment]
> Dependabot _can_ now be added with permissions to push to protected branches, but it _can't_ approve its own PRs. In order to trigger the merge, you still need either a manual approval, or another bot/action (other than dependabot) to approve the PR, even if a PR matches rules in [automerged_updates](https://dependabot.com/docs/config-file/#automerged_updates).

Yep, that's correct! Here's an example GH Action to auto-approve: https://github.com/hmarr/auto-approve-action (going to lock this issue)
